### PR TITLE
tests: Retry test if it fails on framework/Xvfb issue

### PR DESCRIPTION
**Pull request by petervdmeer**
**Feature:** This PR implements a retry mechanism for the CI integration tests if the tests fails on a known framework issue (#4308, #5645)

## Feature Details
If the testframework detects the error-message `SDL message: "No available video device"`, then it will retry to run the test (max 5 times). This SDL errormessage would (usually) not be related to the test itself, but looks like an intermittent issue between endless-sky, Xvfb and Ubuntu 20.04. We should still handle the `SDL message: "No available video device"` error in the future, this PR only acts as a workaround to prevent CI for unrelated PRs from failing on this error-message.

## UI Screenshots
N/A

## Usage Examples
Automatically runs as part of the CI-jobs.

## Testing Done
Tested with succesfull tests, failing tests and tests failing with the specific error-message.

## Performance Impact
N/A

### CODE REVIEW

File: tests/run_tests.sh [hunk 1]
```diff
@@ -25,6 +25,12 @@ function print_graphics_data () {
 }
 
 
+
+function detect_known_issues () {
```
Comment (tehhowch) [inline tests/run_tests.sh:29]: Should we distinguish known issues from _retryable_ issues?

File: tests/run_tests.sh [hunk 2]
```diff
@@ -25,6 +25,12 @@ function print_graphics_data () {
...
+	cat "$1" | grep "SDL message: \"No available video device\""
```
Comment (tehhowch) [inline tests/run_tests.sh:30]: Let's make this easily extendable, for e.g. 0 to N different issues, so a PR to add/remove one is a single line.

**Review by tehhowch — COMMENTED** [review]
Should we break up the test loop into functions? That can make it easier to implement testrunner logic.

File: tests/run_tests.sh (hunk 1 above)
Comment (petervdmeer) [reply tests/run_tests.sh]: Yes, we should. I renamed those to retryable issues.

I believe tests should still fail on known unretryable issues, so I didn't add anything specific for those.

File: tests/run_tests.sh (hunk 2 above)
Comment (petervdmeer) [reply tests/run_tests.sh]: Agreed. I moved the issues list to a separate file.

**Comment by petervdmeer** [comment]
> Should we break up the test loop into functions? That can make it easier to implement testrunner logic.

Yes, I have split off a function for `a single run of a test` and for `a complete test`.

**Review by tehhowch — COMMENTED** [review]
Add `local` where appropriate, then lgtm

File: tests/run_tests.sh [hunk 3]
```diff
@@ -25,6 +27,96 @@ function print_graphics_data () {
...
+# Helper function that executes a single test-run.
+# Parameters:
+# $1 Test(name)
+# Return values:
+# 0 = success
+# 1 = failure
+# 2 = fatal failure (should terminate all testing)
+# 3 = recoverable failure (could retry)
+function run_single_testrun () {
+	TEST="$1"
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply tests/run_tests.sh]: done

**Pull request merged**